src/object/obj_class.c

```diff
@@ -482,3 +496,76 @@ obj_ec_codec_get(daos_oclass_id_t oc_id)
 
 	return NULL;
 }
+
+
+/**
+ * Encode (using ISA-L) a full stripe from the submitted scatter-gather list.
+ *
+ * oid		[IN]		The object id of the object undergoing encode.
+ * sgl		[IN]		The SGL containing the user data.
+ * sg_idx	[IN|OUT]	Index of sg_iov entry in array.
+ * sg_off	[IN|OUT]	Offset into sg_iovs' io_buf.
+ * parity	[IN|OUT]	Struct containing parity buffers.
+ * p_idx	[IN]		Index into parity p_bufs array.
+ */
+int
+obj_encode_full_stripe(daos_obj_id_t oid, daos_sg_list_t *sgl, uint32_t *sg_idx,
+		       size_t *sg_off, struct obj_ec_parity *parity, int p_idx)
+{
+	struct obj_ec_codec		*codec = obj_ec_codec_get(
+							daos_obj_id2class(oid));
+	struct daos_oclass_attr		*oca = daos_oclass_attr_find(oid);
+	unsigned int			 len = oca->u.ec.e_len;
+	unsigned int			 k = oca->u.ec.e_k;
+	unsigned int			 p = oca->u.ec.e_p;
+	unsigned char			*data[k];
+	unsigned char			*ldata[k];
+	int				 i, lcnt = 0;
+	int				 rc = 0;
+
+	for (i = 0; i < k; i++)
+		if (sgl->sg_iovs[*sg_idx].iov_len - *sg_off >= len) {
+			unsigned char *from =
+				(unsigned char *)sgl->sg_iovs[*sg_idx].iov_buf;
+
+			data[i] = &from[*sg_off];
+			*sg_off += len;
+			if (*sg_off == sgl->sg_iovs[*sg_idx].iov_len) {
+				*sg_off = 0;
+				(*sg_idx)++;
+			}
+		} else {
+			int cp_cnt = 0;
+
+			D_ALLOC(ldata[lcnt], len);
+			if (ldata[lcnt] == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			while (cp_cnt < len) {
+				int cp_amt =
+					sgl->sg_iovs[*sg_idx].iov_len-*sg_off <
+					len - cp_cnt ?
+					sgl->sg_iovs[*sg_idx].iov_len-*sg_off :
+					len - cp_cnt;
+				unsigned char *from =
+					sgl->sg_iovs[*sg_idx].iov_buf;
+
+				memcpy(&ldata[lcnt][cp_cnt], &from[*sg_off],
+				       cp_amt);
+				if (sgl->sg_iovs[*sg_idx].iov_len - *sg_off <
+					len - cp_cnt) {
+					*sg_off = 0;
+					(*sg_idx)++;
+				} else
+					*sg_off += cp_amt;
+				cp_cnt += cp_amt;
+			}
+			data[i] = ldata[lcnt++];
+		}
+
+	ec_encode_data(len, k, p, codec->ec_gftbls, data,
+		       &parity->p_bufs[p_idx]);
```

Comment:
The len variable represents the cell size. This will never be smaller than 16KB, and the current lower bound is 32KB.  I'm not a big fan of using this variable name (len) for the cell size (I think its too generic), but Xuezhao wants to use the same names as in the ISA-L. If we were to use a value smaller than the cell size, we could perhaps avoid the allocation, but we would be challenged to determine the size that would allow contiguous buffers that align across the stripe (NP-complete, I suspect)._ You're right if we do an encode length of one, we would avoid the need the allocate, but we'd execute the ISA-L encode function cell-size number of times (and not take advantage of the SSE wide instructions). We have an EC object class with a 1MB cell size, so that would be a lot of calls to encode.  The memcpy looks bad, but it's important to remember that it only is invoked when the SGL is fragmented into iov_bufs that are smaller than the cell size. The common case will be no memcpy.

Reply:
if you think the name e_len is not good, I'm fine to rename it, for example e_cell_size or e_cs;

Reply:
oh, I do not mean you should call ec_encode with 1 byte length. I mean once some data[i] reach to the end of one sg_iovs[i].iov_buf, you can just call ec_encode(), instead of allocate ldata and memcpy. 

So in most cases, it still only call 1 ec_encode, but if the user buffer include a lot of small fragmented sgls, then you may need call ec_encode many times indeed. Btw: as I see from the code I see, invoking ec_encode_data() multiple times with small buffers seems not much worse than only single ec_encode_data() with big buffer.  Unless there are hardware support here, do we?

Reply:
So it's not as simple as looking at a single iov_buf.  You need to get k of them that are aligned across the stripe.  When you fill the data buffers to the cell length, then the alignment is guaranteed.  But if you try to do sub-stripes that avoid any copies, it's far more complex.  Also, Xuezhao says that ISL-L doesn't support this (I haven't looked into it).

Reply:
Hmm, it seems to me. it can be done simply as

1. locate the start pointer for each data cells over the sgl.
2. Then do the loop like
       while (encode_length < cell_size) {
                 /* find the shortest but_size to the end of sgl_iov[i].iov_buf among all data cells */
                 current_size = find_shortest_sizes_over_sgls(datas[], cell_cnt, *iov_offs[], ....);
                
                ec_encode(current_size, datas, ....);

                encode_length += current_size;

               /* Update the shortest data[i] pointer */
      }
Maybe I miss sth here? 
I am fine to use allocate and memcpy if this is only for client side. But we may need do a lot encoding during server rebuild, so I am a bit concern about these overhead. Maybe we can add a wrapper ec_encode_sgl_xxx() to use sgl as input for now, and optimize the internal of this function later.  Anyway this is just a suggestion.

Btw: xuezhao, what do you mean "ISL-L doesn't support this "?  you mean ISL-L require the encode buffer has to == data cell size?  I did not see that, or I miss sth? Thanks.

Reply:
please see base version encode:
301 void ec_encode_data_base(int len, int srcs, int dests, unsigned char *v,        
302 >       >       >        unsigned char **src, unsigned char **dest)             
303 {                                                                               
304 >       int i, j, l;                                                            
305 >       unsigned char s;                                                        
306                                                                                 
307 >       for (l = 0; l < dests; l++) {                                           
308 >       >       for (i = 0; i < len; i++) {                                     
309 >       >       >       s = 0;                                                  
310 >       >       >       for (j = 0; j < srcs; j++)                              
311 >       >       >       >       s ^= gf_mul(src[j][i], v[j * 32 + l * srcs * 32 + 1]);
312                                                                                 
313 >       >       >       dest[l][i] = s;                                         
314 >       >       }                                                               
315 >       }                                                                       
316 }                    

How could you do? we can change the src[j][i] to be (sgl[sg_idx].iov + off++) to avoid memcpy, but many time need to check if need to advance to next sg_idx. this possible with less performance compared to just with memcpy, and for most case we don't need memcpy.
Node that for EC encode, the parity code is calculated in two-dimension matrix manner, not one-dimension stream data.

Moreover, there is SSE/AVX version implementation for ec_encode_data() with much more better performance than basic version, it it hard to change the ASM code for now.

I think we can just based on current en_encode_data() interface.

Reply:
I think you are right, encode_data needs the offset from start, and it always start from 0. sigh. Ok. Sorry for the noise.